src/Confluent.Kafka/Headers.cs

```diff
@@ -0,0 +1,100 @@
+// Copyright 2018 Confluent Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Refer to LICENSE for more information.
+
+using System;
+using System.Collections.Generic;
+
+
+namespace Confluent.Kafka
+{
+    /// <summary>
+    ///     A collection of Kafka message headers.
+    /// </summary>
+    public class Headers : List<KeyValuePair<string, byte[]>>
+    {
+        /// <summary>
+        ///     Append a new header to the collection.
+        /// </summary>
+        /// <param name="key">
+        ///     The header key.
+        /// </param>
+        /// <param name="val">
+        ///     The header value.
+        /// </param>
+        public void Add(string key, byte[] val)
+        {
+            if (key == null) 
```

Comment:
This is not a failsafe check. The user of this class can still add a null value by invoking List.Add (which is not virtual, so it's not a good idea to re-implement it in this class). However I don't think it's a big enough deal to implement this class more rigorously - the produce calls all fail with a null argument exception if a header key is null.